doc/man/monitest.1
.TH "monitest" 1 "2003-04-02" "libggi-current" GGI
.SH NAME
\fBmonitest\fR : Monitor test program
.SH SYNOPSIS
.nb
.nf
monitest [modestring]
.fi

.SH DESCRIPTION
\fBmonitest\fR is intended to test ggi drivers (during development) and
the optical quality of monitors (later). Don't be too disappointed if
you find a weakness in your monitor, mine is probably worse than yours
...

The main feature is a test screen like broadcast in former days before
24 hour TV programs. It is used to test geometry and resolution of the
monitor.

Also included are several test screens for testing moiree effects, the
horizontal and vertical screen resolution, and colour convergence.

You will be able to switch resolution on the fly, to find out
interactively what modes the driver supports and how much of this your
card/monitor can do with acceptable quality.
.SH OPTIONS
.TP
\fBmodestring\fR
Specify the mode to use

.PP
.SH PATTERN DESCRIPTIONS
.SS TESTPATTERN
The basic grid is white on black and has sixteen times twelve
fields. On a tube with 4:3 ratio each one should be square, even if
the pixel size of the screen does not have 4:3 ratio. The lines are
one pixel wide. They should be straight, even in the corners, and not
have coloured borders.

There is a big circle in the middle and smaller circles in each
corner.  They are round pixelwise, so they should be circles if the
screen size ratio (width:height of visible area) is equal to the pixel
ratio. It usually should be 4:3. So the best sizes for testing are
320x240, 640x480, 800x600 and up.

In each corner there is one box with vertical stripes. These are one
pixel wide, with one pixel distance, so you get maximum signal
frequency and can see how well your monitor and video card handle the
dotclock.

The middle field has eight solid blocks with the eight colours, i.e
all combinations of the red, green and blue signals turned on and
off. Below it there are four fields with vertical stripes as in the
corners, but in white and the three basic colours red, green and
blue. Below it there is a bar with these four colours red, blue, green
and white blending from full intensity (left) to zero intensity
(right), i.e. black.

In the middle the current resolution is printed. Maybe horizontal and
vertical frequency will be printed too, if I can get the information,
which is not (yet) implemented in the LibGGI API.
.SS CONVERGENCE
Convergence means how well the red, green and blue picture are
aligned.

This is tested by painting a grid of red, green and blue + signs. They
should be aligned properly where they touch. Usually they don't...

There are four patterns like this, rotating the colours. Press space
to switch forward, or press q to quit anytime.
.SS RESOLUTION
Once again there are several screens, press space to step thru them.

Vertical white stripes with width 1, 2, 3 and 4 pixels. See what the
highest dotclock is the monitor can handle.

Horizontal stripes with width 1 and 2 pixels. See how well the scan
lines are separated.

Three stars of black lines on white, with a width and space (at the
sceen borders) of 1 and 1, 1 and 5, 2 and 10 respectively. Watch for
colour changes, and once again you can see the maximum frequency your
monitor can do.

Vertical bars in red, blue, green and white, with the width of 4, 3,
2, 1, 2, 3 and 4 pixels for bar and space. See whether there is a
difference in resolution between the colours. And watch, again, for
the alignment.
.SS MOIREE
I don't know whether these tests work, my monitor is rather good in
this respect. Please gimme some feedback!

If there is interference between the monitor mask and a grid displayed
on the monitor, a change of colours can be seen, and is sometimes very
annoying. There are three tests present, each one comes in the four
colour combinations black, red, green and blue on white background :
vertical stripes, one dot wide, with one dot space; white dots on
colour ground, spaced two and two (run testscreen with a really low
resolution to see what I mean :-); a chessboard.
.SS FLAT PANEL TEST
This test allows you to drag a coloured rectangle around, looking for
pixels that are always on or off, which is, as far as I know, the most
common failure of flat panels.

The rectange is moved using the mouse or the arrow keys.

The color of the rectange can be changed by pressing the primary mouse
button (usually the left one) or &lt;Space&gt;, cycling through
(black, red, green, blue, white) or by pressing a number between 0 and
4 or the first letter of the colour (b is blue).

Dragging with the second button pressed changes the size of the
rectangle. Every other key terminates this test.
.SH BUGS AND LIMITATIONS
.IP \(bu 4
If you switch depth, the program might crash badly. This will be
solved once I figure out mode checking or using a target that
(opposed to the X targets) supports that.
.PP
.SH AUTHOR
\fBmonitest\fR was written by Hartmut Niemann.


NEWS
NEWS - list of user-visible changes between releases of LibGGI

New in 2:1:0 (2.0.1): 2001-08-26
* More documentation updates.
* Prelimiary support for using DirectFB drivers under display-fbdev
* Many corrections/improvements to build files
* Started move towards using libtool versioning, as interface is stable.


New in 2.0: 2001-08-11
* Huge documentation updates.
* svgalib-target
  - added frame support
  - misc bug-fixes
^L


New in 2.0beta4: 2001-07-10
* Misc bugfixes (see ChangeLog).
* Huge documentation updates.
* finished support for reporting physical screen sizes.
* splitted libggimisc off
* New API: ggiDetachInput()
* fbdev-target
  - Acceleration of Matrox G400 (matrox fb)
* directX-target
  - Update
  - Highly experimental target for Win ... Use at your own risk !
* ipc-target
  - New target for interprocess communication (experimental)
^L


New in 2.0beta3: 2000-12-18
* Misc bugfixes (see ChangeLog).
* Minor documentation updates.
* Major rewrite of the internal sublib handling.
* Some fixes for PowerPC compatibility.
* preliminary support for reporting physical screen sizes.
* Made all headers C++ compatible.

* lcd823-target
  - Target for the Motorola MPC823 builtin LCD controler, using
    the /dev/lcd823 Linux kernel device.
* fbdev-target
  - Better detection of kgicon-type fb-devices and old fb implementations.
  - Better usage of fbdev mode infos in fb.modes
  - Better acceleration for the Matrox fb.
* glide-target
  - Added directbuffer support.
* aalib-target
  - Better mode handling.
* X-target
  - Propagated X dpi values to give physical screen dimensions.
  - Various fixes to properly init several values.
* tele-target
  - Updated much to make it work again.
* vgl-target
  - New target for FreeBSD's VGL library.
* directX-target
  - Highly experimental target for Win ... Use at your own risk !



New in 2.0beta2.1: 1999-05-17;
* Misc bugfixes (see ChangeLog).
* Minor documentation updates.
* Autoconf now checks for headers as well as libraries for Glide, AAlib and
  SVGAlib.
* Should now compile on Linux libc5 and other systems which doesn't define
  MAP_FAILED.

* cube3d-demo
  - Fixed to work on GT_32BIT visuals.
* demo-demo
  - Better teach how to properly use ASYNC/SYNC mode.

* fbdev-target
  - Now supports acceleration on matroxfb. Should work on all supported cards.
  - Read /etc/fb.modes to get timing information - mode switching now
    works for non-KGIcon drivers too.
  - Fixed bug when running on a new VT (like XGGI does by default).
  - Don't try to change the palette when the application is switched away.
  - Now compiles on 2.0 kernels - someone with the proper hardware please
    test whether it actually works.
* svgalib-target
  - Major updates and bugfixes.
  - Now supports running in the background using the GGICMD_NOHALT_ON_UNMPAP
    standard API.
  - Console switching now works, but does _not_ use SIGUSR* - you can now use
    the SVGAlib target with pthreads!



New in 2.0beta2: 1999-04-01;
* Portability fixes for non-Linux platforms.
* Lots of bugfixes.
* Documentaion updates.
* Optimizations of the linear-* drawing libraries - now software drawing,
  ggiCrossBlit() and ggiPutBox() is much faster.
* Linux VT-switching now works with pthreads.
* New environment variable GGI_NEWVT which will cause Linux-console based
  targets to open a new VT instead of running on the current one.
* API for requesting Linux-console based targets not to halt when switching
  to another VT.
* New API functions ggiSetColorfulPalette(), ggiFlushRegion(),
  ggiResourceAcquire(), ggiResourceRelease(), ggiResourceMustAcquire(),
  ggiResourceFastAcquire(), ggiResourceFastRelease(), ggiEventsQueued(),
  ggiEventSend() and ggiEventSelect().
* ggiGetSelectFdSet() removed.
* The config file is now stored in /usr/local/etc/ggi by default.
* Support for setting the mansync framerate.
* New configure option --with-uninst-gii, which makes it possible to
  build LibGGI against an uninstalled copy of LibGII.
* All targets should now be compatible with Linux devfs systems.
* Support for targets using asynchronous acceleration.
* Big improvements in the event handling code, see the LibGII docs and NEWS
  file for info.

* stars-demo
  - Demonstrate how to runtime switch between X and DGA targets with help
    from the application.
* cube3d-demo
  - New demo which lets you run one to six arbitrary LibGGI applications
    on a rotating, zoomable, cube, which is displayed on any LibGGI target.

* X-target
  - Handle expose events internally.
* X/Xlib-targets
  - Much improved system for matching LibGGI graphtypes with X-visual types.
  - Support for StaticColor X-visuals.
  - Support for running in an already existing window or in the root window.
  - Don't close and reopen the window when changing modes.
* DGA-target
  - Use our own client-side implementation of XFree86DGA to get better
    stability and support for optionally mapping the video memory from
    a framebuffer device (= no root privs required for DGA).
  - Support for acceleration on X servers supporting DGA 1.1.
* fbdev-target
  - Fixed fatal bug which would cause segfault when stride != virtual x.
* glide-target
  - Improved GGI_AUTO handling, support for more resolutions.
  - Support for more than one 3DFX card.
  - Restrained refreshrates so you doesn't damage your monitor.
* svgalib-target
  - 32-bit modes should now work
  - "return key causes quit" problem fixed.
* vcsa-target
  - Much improved color mapping algorithm.
  - "-shade" option added, and made shading NOT be the default since it
     screws up the colormapping for normal text.
* file-target
  - Support for saving movies of running GGI applications.



programs/util/monitest/monitest.c
/* $Id: monitest.c,v 1.6 2004/09/08 17:51:03 cegger Exp $
******************************************************************************

   Monitor test pattern generator

   Written in 1998 by Hartmut Niemann

   This software is placed in the public domain and can be used freely
   for any purpose. It comes without any kind of warranty, either
   expressed or implied, including, but not limited to the implied
   warranties of merchantability or fitness for a particular purpose.
   Use it at your own risk. the author is not responsible for any damage
   or consequences raised by use or inability to use this program.
 
******************************************************************************

   This program creates a test pattern like ctmon3 and the
   Nokia montest program created or is/was broadcast on tv
   during program breaks, when there was no 24h program.

   The second part is a convergence test researched from the
   Nokia program, to check whether red, green and blue beams
   are adjusted properly.

   for a first glance: run
	monitest 640x480 
   or
	monitest

   Very much is copied from ./demo.c
   The text output would require about 400x300 dots, but it works
   with less too.

******************************************************************************
*/

#include <ggi/ggi.h>

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

#include "monitest.h"
#include "ggitext.h"
#include "menu.h"

#include "drawlib.h"

#ifdef HAVE_KGIMON_H
#include "kgitune.h"
#else
#define is_kgicondev(x) (0)
#define kgitune(x)	do{}while(0)
#endif

static ggi_visual_t vis;
static int kgidevice;

ggi_pixel white, black, red, green, blue, yellow, magenta, cyan;

static void usage(const char *prog)
{
	fprintf(stderr, "Usage:\n\n"
		"%s <bpp> <xsize> <ysize>\n\n"
		"Default: %s 8 320 200\n", prog, prog);
	exit(1);
}

int waitabit(ggi_visual_t _vis)
{
	int key;
	key = ggiGetc(_vis);

	if (toupper(key) == 'Q') {
		/* Q pressed */
		ggiClose(_vis);
		ggiExit();
		exit(1);
	} else if (key == GIIUC_Escape || toupper(key) == 'B') {
		return 1;
	}
	return 0;
}


static void setcolors(void)
{
	/* just set the needed colour names */
	ggi_color col;

	col.r = 0xFFFF;
	col.g = 0xFFFF;
	col.b = 0xFFFF;
	white = ggiMapColor(vis, &col);

	col.r = 0xFFFF;
	col.g = 0xFFFF;
	col.b = 0x0000;
	yellow = ggiMapColor(vis, &col);

	col.r = 0xFFFF;
	col.g = 0x0000;
	col.b = 0xFFFF;
	magenta = ggiMapColor(vis, &col);

	col.r = 0xFFFF;
	col.g = 0x0000;
	col.b = 0x0000;
	red = ggiMapColor(vis, &col);

	col.r = 0x0000;
	col.g = 0xFFFF;
	col.b = 0xFFFF;
	cyan = ggiMapColor(vis, &col);

	col.r = 0x0000;
	col.g = 0xFFFF;
	col.b = 0x0000;
	green = ggiMapColor(vis, &col);

	col.r = 0x0000;
	col.g = 0x0000;
	col.b = 0xFFFF;
	blue = ggiMapColor(vis, &col);

	col.r = 0x0000;
	col.g = 0x0000;
	col.b = 0x0000;
	black = ggiMapColor(vis, &col);
}


static void convergence(ggi_visual_t _vis)
{
#define PARTSHOR 16
#define PARTSVERT 12
#define PARTX(x) (signed)((x)*(xmax-1)/PARTSHOR)
#define PARTWIDTH (signed)(xmax/PARTSHOR)
#define PARTY(y) (signed)((y)*(ymax-1)/PARTSVERT)
#define PARTHEIGHT (signed)(ymax/PARTSVERT)

	unsigned int i, j, k, xmax, ymax;
	unsigned int f[3];
	ggi_mode currmode;

	ggiGetMode(_vis, &currmode);
	xmax = currmode.visible.x;
	ymax = currmode.visible.y;

	f[0] = red;
	f[1] = green;
	f[2] = blue;

	for (k = 0; k < 4; k++) {
		ggiSetGCForeground(_vis, black);
		ggiFillscreen(_vis);
		for (i = 0; i <= PARTSHOR; i++) {
			for (j = 0; j <= PARTSVERT; j++) {
				ggiSetGCForeground(_vis,
						   f[(i + j + k) % 3]);
				ggiDrawHLine(_vis, PARTX(i), PARTY(j),
					     PARTWIDTH / 2);
				ggiDrawVLine(_vis, PARTX(i), PARTY(j),
					     PARTHEIGHT / 2);
				ggiDrawHLine(_vis,
					     PARTX(i) - PARTWIDTH / 2,
					     PARTY(j), PARTWIDTH / 2);
				ggiDrawVLine(_vis, PARTX(i),
					     PARTY(j) - PARTHEIGHT / 2,
					     PARTHEIGHT / 2);
			}
		}
		ggiFlush(_vis);
		if (waitabit(_vis))
			return;
	}
}


static void moiree(ggi_visual_t _vis)
{
	int xmax, ymax;
	ggi_mode currmode;

	ggiGetMode(_vis, &currmode);
	xmax = currmode.visible.x;
	ymax = currmode.visible.y;
	stripevert(_vis, 0, 0, xmax - 1, ymax - 1,
		   black, white, 1);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	stripevert(_vis, 0, 0, xmax - 1, ymax - 1,
		   red, white, 1);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	stripevert(_vis, 0, 0, xmax - 1, ymax - 1,
		   green, white, 1);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	stripevert(_vis, 0, 0, xmax - 1, ymax - 1,
		   blue, white, 1);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;

	dotone(_vis, 0, 0, xmax - 1, ymax - 1, black, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	dotone(_vis, 0, 0, xmax - 1, ymax - 1, red, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	dotone(_vis, 0, 0, xmax - 1, ymax - 1, green, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	dotone(_vis, 0, 0, xmax - 1, ymax - 1, blue,
	       white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;

	chessboardone(_vis, 0, 0, xmax - 1, ymax - 1, black, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	chessboardone(_vis, 0, 0, xmax - 1, ymax - 1, red, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	chessboardone(_vis, 0, 0, xmax - 1, ymax - 1, green, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
	chessboardone(_vis, 0, 0, xmax - 1, ymax - 1, blue, white);
	ggiFlush(_vis);
	if (waitabit(_vis))
		return;
}

#if 0	/* defined but not used */
char *helptext = {
	"GGI screntest program               \n"
	    "(c) H. Niemann, $Id: monitest.c,v 1.6 2004/09/08 17:51:03 cegger Exp $               \n"
	    "h:   this help screen               \n"
	    "q:   quit this testscreen           \n" ""
};

static int help(void)
{
	ggi_mode currmode;

	ggiGetMode(vis, &currmode);

	ggiGraphTextLongPuts(vis, currmode.visible.x / 2,
			     currmode.visible.y / 2, 0, 0, GGI_TEXT_CENTER,
			     helptext);
	waitabit(vis);
	return 0;
}
#endif


static const ggi_coord resolutions[] = {
	{320, 200},
	{320, 240},
	{320, 350},		/* EGA text */
	{360, 400},		/* VGA text */
	{400, 300},
	{480, 300},
	{512, 384},
	{640, 200},		/* EGA */
	{640, 350},		/* EGA */
	{640, 400},
	{640, 480},
	{720, 350},		/* MDA text */
	{720, 400},		/* VGA text */
	{800, 600},
	{1024, 768},
	{1152, 864},
	{1280, 1024},
	{1600, 1200},
	{0, 0}			/* End mark!! */
};

static const ggi_graphtype graphtypes[] = {
	/* can't handle textmode yet */
	GT_1BIT,		/*  1 bpp graphics              */
	GT_4BIT,		/*  4 bpp graphics              */
	GT_8BIT,		/*  8 bpp graphics              */
	GT_15BIT,		/* 15 bpp graphics              */
	GT_16BIT,		/* 16 bpp graphics              */
	GT_24BIT,		/* 24 bpp graphics              */
	GT_32BIT,		/* 24 bpp word aligned          */
	GT_INVALID
};

static int resindex = -1;		/* 'base' resolution of the mode set */
static int gtindex = -1;		/* graphtype, i.e. bit depth         */
static int xres = GGI_AUTO;		/* visual resolution, might be different */
static int yres = GGI_AUTO;		/* from the mode list */

static int guessmode(void)
{
/* try to guess the mode you're in and set resindex and gtindex  */
/* done on startup.                                              */
/* I could set up a certain mode at startup, but I don't want to */

	ggi_mode currmode;

	int i;

	ggiGetMode(vis, &currmode);
	fprintf(stderr, "Current mode is %dx%d %u/%u.\n",
		currmode.visible.x, currmode.visible.y,
		GT_DEPTH(currmode.graphtype), GT_SIZE(currmode.graphtype));


	xres = currmode.visible.x;
	yres = currmode.visible.y;

	/* find most possible resolution */
	resindex = 0;		/* if everything else fails: assume first mode:320x200 */
	i = 0;
	while (resolutions[i].x != 0) {
		/* for now: only test exact match */
		if ((xres == resolutions[i].x)
		    && (yres == resolutions[i].y)) {
			resindex = i;
			fprintf(stderr, "Detected resolution %d: %dx%d.\n",
				resindex,
				resolutions[resindex].x,
				resolutions[resindex].y);
		}
		i++;
	}
	/* find mode you're in */
	gtindex = -1;
	i = 0;
	while (graphtypes[i] != GT_INVALID) {
		if (graphtypes[i] == currmode.graphtype) {
			gtindex = i;
			fprintf(stderr, "Detected graphtype %u/%u\n",
				GT_DEPTH(graphtypes[gtindex]),
				GT_SIZE(graphtypes[gtindex]));
			break;
		}
		i++;
	}
	if (gtindex == -1) {
		return -1;
	}
	return 0;
}


static int changeresmenu(void)
{
	struct menu cm;
	static int _select = 0;

	ggi_mode suggmode;

	char s[32];
	char nextmodeline[100];
	char prevmodeline[100];

	char nextgtline[100];
	char prevgtline[100];

	char bottom[100];

	int nextresindex, prevresindex;
	int nextgtindex, prevgtindex;

	nextresindex = prevresindex = 0;
	nextgtindex = prevgtindex = 0;

	default_menu(&cm, vis);

	cm.w.title = " Change Screen resolution ";

	cm.lastentry = 8;
	cm.entry[0].text = "1 Increase X";
	cm.entry[1].text = "2 Decrease X";
	cm.entry[2].text = "3 Increase Y";
	cm.entry[3].text = "4 Decrease Y";
	cm.entry[8].text = "9 Back";

	sprintf(bottom, "  ");

	for (;;) {

		nextresindex = resindex + 1;
		if (resolutions[nextresindex].x == 0) {
			nextresindex--;
		};
		prevresindex = resindex - 1;
		if (prevresindex < 0) {
			prevresindex = 0;
		}

		nextgtindex = gtindex + 1;
		if (graphtypes[nextgtindex] == GT_INVALID) {
			nextgtindex--;
		}
		prevgtindex = gtindex - 1;
		if (prevgtindex < 0) {
			prevgtindex = 0;
		}
		sprintf(cm.entry[4].text = nextgtline,
			"5 Increase depth: %u/%u",
			GT_DEPTH(graphtypes[nextgtindex]),
			GT_SIZE(graphtypes[nextgtindex]));
		sprintf(cm.entry[5].text = prevgtline,
			"6 Decrease depth: %u/%u",
			GT_DEPTH(graphtypes[prevgtindex]),
			GT_SIZE(graphtypes[prevgtindex]));
		sprintf(cm.entry[6].text =
			nextmodeline, "7 next res.:  %dx%d",
			resolutions[nextresindex].x,
			resolutions[nextresindex].y);
		sprintf(cm.entry[7].text =
			prevmodeline, "8 prev. res.: %dx%d",
			resolutions[prevresindex].x,
			resolutions[prevresindex].y);


		sprintf(s, "current: %4dx%3dx%u/%u", xres, yres,
			GT_DEPTH(graphtypes[gtindex]),
			GT_SIZE(graphtypes[gtindex]));

		cm.toptext = s;
		cm.bottomtext = bottom;

		calculate_menu(&cm);
		center_menu(&cm);

		/*sleep(1); */
		testpattern(vis);
		ggiFlush(vis);
		switch (_select = do_menu(&cm, _select)) {
		case 0:
			xres += 1;
			break;
		case 1:
			xres -= 1;
			break;
		case 2:
			yres += 1;
			break;
		case 3:
			yres -= 1;
			break;
		case 4:
			gtindex = nextgtindex;
			break;
		case 5:
			gtindex = prevgtindex;
			break;
		case 6:
			resindex = nextresindex;
			xres = resolutions[resindex].x;
			yres = resolutions[resindex].y;
			break;
		case 7:
			resindex = prevresindex;
			xres = resolutions[resindex].x;
			yres = resolutions[resindex].y;
			break;
		case 8:
		case -1:
			testpattern(vis);	/* clear menu */
			return (0);
		default:
			ggiPanic("Internal error, wrong menu selection");
		}
		/* try mode */
		if (ggiCheckSimpleMode
		    (vis, xres, yres, 1, graphtypes[gtindex],
		     &suggmode) != 0) {
			/* failed */
			sprintf(bottom, "mode set failed");
		} else {
			if (ggiSetSimpleMode(vis, xres, yres, 1,
					     graphtypes[gtindex]) != 0) {
				fprintf(stderr, "Set may not fail!!\n");
			} else {
				sprintf(bottom, "mode set succeeded");
			}
			/*sleep(1); */
			/*ggiFlush(vis); */
		}
		setcolors();	/* necessary when depth changed */

	}
	return 0; /* never get here */
}



static int mainmenu(void)
{
	struct menu mainm;
	static int _select = 0;

	default_menu(&mainm, vis);
	mainm.w.title = " * Main menu * ";

	mainm.entry[0].text = "1 Geometry and Colors";
	mainm.entry[1].text = "2 Convergence";
	mainm.entry[2].text = "3 Resolution";
	mainm.entry[3].text = "4 Moiree";
	mainm.entry[4].text = "5 Change Resolution";
	mainm.entry[5].text = "6 Flat panel check";
	if (kgidevice) {
		mainm.entry[6].text = "7 Tune KGIcon monitor timings";
		mainm.entry[7].text = "8 Exit";
		mainm.lastentry = 7;
	} else {
		mainm.entry[6].text = "7 Exit";
		mainm.lastentry = 6;
	}

	mainm.toptext = "Please choose with arrows";
	mainm.bottomtext = "and press <return>";

	calculate_menu(&mainm);

	for (;;) {
		center_menu(&mainm);
		/* if resolution changes, we must recenter */
		switch (_select = do_menu(&mainm, _select)) {
		case 0:
			testpattern(vis);
			waitabit(vis);
			break;
		case 1:
			convergence(vis);
			break;
		case 2:
			resolution(vis);
			break;
		case 3:
			moiree(vis);
			break;
		case 4:
			changeresmenu();
			break;
		case 5:
			flatpanel(vis);
			break;
		case 6:
			if (kgidevice) {
				testpattern(vis);
				kgitune(vis);
			} else {
				return 0;
			}
			break;
		case 7:
		case -1:
			return 0;
			break; /* never get here */
		default:
			ggiPanic("Internal error, wrong menu selection");
		}
	}
	return 0; /* never get here */
}


int main(int argc, char **argv)
{
	const char *prog = argv[0];
	ggi_mode mo;

	if (argc == 1) {
		ggiParseMode("", &mo);
	} else if (argc == 2) {
		ggiParseMode(argv[1], &mo);
	} else {
		usage(prog);
		return 1;
	}


	if (ggiInit() < 0) {
		fprintf(stderr, "unable to initialize LibGGI, exiting.\n");
		exit(1);
	}

	vis = ggiOpen(NULL);	/* Null gives the default visual */
	if (vis == NULL) {
		ggiPanic("unable to open default visual, exiting.\n");
	}

	ggiSetFlags(vis, GGIFLAG_ASYNC);

	printf("Trying mode ");
	ggiPrintMode(&mo);
	printf("\n");

	ggiCheckMode(vis, &mo);
	/*ggiCheckTextMode(vis,80,25,80,25,9,14,GT_TEXT16,&mo,NULL); */

	printf("Suggested mode ");
	ggiPrintMode(&mo);
	printf("\n");

	if (ggiSetMode(vis, &mo) != 0) {
		ggiPanic("unable to set any mode at all, exiting!\n");
		return 2;
	}

	kgidevice = is_kgicondev(vis);

	setcolors();

	if (guessmode() != 0) {
		fprintf(stderr,
			"Warning: Could not guess mode, probably not a standard mode\n");
	}

	testpattern(vis);

	mainmenu();

	ggiClose(vis);

	ggiExit();

	return 0;
}
